rasa/core/actions/two_stage_fallback.py

```diff
@@ -0,0 +1,197 @@
+import copy
+import time
+from typing import List, Text, Optional
+
+from rasa.constants import DEFAULT_NLU_FALLBACK_INTENT_NAME
+from rasa.core.actions import action
+from rasa.core.actions.action import (
+    ACTION_DEFAULT_ASK_AFFIRMATION_NAME,
+    ACTION_LISTEN_NAME,
+    ACTION_DEFAULT_FALLBACK_NAME,
+    ACTION_DEFAULT_ASK_REPHRASE_NAME,
+)
+from rasa.core.actions.loop import LoopAction
+from rasa.core.channels import OutputChannel
+from rasa.core.constants import USER_INTENT_OUT_OF_SCOPE
+from rasa.core.domain import Domain
+from rasa.core.events import (
+    Event,
+    UserUtteranceReverted,
+    ActionExecuted,
+    UserUttered,
+    Form,
+)
+from rasa.core.nlg import NaturalLanguageGenerator
+from rasa.core.trackers import DialogueStateTracker
+from rasa.utils.endpoints import EndpointConfig
+
+ACTION_TWO_STAGE_FALLBACK_NAME = "two_stage_fallback"
+
+
+class TwoStageFallbackAction(LoopAction):
+    def __init__(self, action_endpoint: Optional[EndpointConfig] = None) -> None:
+        self._action_endpoint = action_endpoint
+
+    def name(self) -> Text:
+        return ACTION_TWO_STAGE_FALLBACK_NAME
+
+    async def do(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> List[Event]:
+        if _user_should_affirm(tracker, events_so_far):
```

Comment:
why is it not a classmetod of this class?

Reply:
I'm following the [Google Styleguide on this one](http://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators). I used to do it differently, but I think wiser people than me thought this through 😀